### ISSUE TITLE
sqlite3_key() has a potential danger

### ISSUE THREAD

**Issue by ComeOnEric** [sqlcipher/sqlcipher#164]
HI.
#### sqlcipher version: v3.3.0

when i call the function sqlite3_key() to set key, the function will call sqlcipher_codec_ctx_set_cipher() to set cipher, in the function sqlcipher_codec_ctx_set_cipher():
  `c_ctx->provider->set_cipher(c_ctx->provider_ctx, cipher_name);`
set_cipher() will call sqlcipher_openssl_set_cipher:

```
static int sqlcipher_openssl_set_cipher(void *ctx, const char *cipher_name) {
  openssl_ctx *o_ctx = (openssl_ctx *)ctx;
  o_ctx->evp_cipher = (EVP_CIPHER *) EVP_get_cipherbyname(cipher_name);
  return SQLITE_OK;
}
```

if EVP_get_cipherbyname() fail, then c_ctx->provider_ctx will be null. the following code:
  `c_ctx->key_sz = c_ctx->provider->get_key_sz(c_ctx->provider_ctx);`
will call sqlcipher_openssl_get_key_sz():

```
static int sqlcipher_openssl_get_key_sz(void *ctx) {
    return EVP_CIPHER_key_length(((openssl_ctx *)ctx)->evp_cipher);
}
```

in the EVP_CIPHER_key_length(), it does not check the pointer "ctx" whether is null, and direct operate the pointer. 

```
int EVP_CIPHER_key_length(const EVP_CIPHER *cipher)
{
    return cipher->key_len;
}
```

Because the pointer "cipher" is null ,  the program will  collapse.

I know in the v3.3.1, you modify the function "sqlcipher_openssl_set_cipher" to avoid the problem :

```
static int sqlcipher_openssl_set_cipher(void *ctx, const char *cipher_name) {
  openssl_ctx *o_ctx = (openssl_ctx *)ctx;
  EVP_CIPHER* cipher = (EVP_CIPHER *) EVP_get_cipherbyname(cipher_name);
  if(cipher != NULL) {
    o_ctx->evp_cipher = cipher;
  }
  return cipher != NULL ? SQLITE_OK : SQLITE_ERROR;
}
```

but in function "sqlcipher_codec_ctx_set_cipher", still does not check the return value of "set_cipher()"
so the problem is not solved.
thanks.

**Comment by developernotes** [comment]
Hi @ComeOnEric 

Thank you for sharing your review. We are looking into a change that would invalidate the pager in the scenario you propose above. Thanks!

**Comment by brodybits** [comment]
I would also love to see this one fixed. While I don't think we can categorize this as a security issue in case the OpenSSL library is built and used correctly I would like to see SQLCipher block the data storage in case this goes wrong.

**Comment by sjlombardo** [comment]
This issue was resolved with commit c1f7fe783c25058517c6c90c5d38e198c8d32fa2.

**Comment by brodybits** [comment]
> This issue was resolved with commit c1f7fe7.

... already part of SQLCipher 3.4.1 and 3.4.2, or am I mistaken somehow?

Thanks again to the SQLCipher team for your efforts and dedication!

**Comment by sjlombardo** [comment]
Hi @brodybits yes, you absolutely correct, this fix was included in the 3.4.1 and 3.4.2 releases. I just noticed this old ticket sitting out here and wanted to close the loop. Thanks for keeping an eye on things as well!
